### PR TITLE
Validate attribute types and use more specific types for storing them

### PR DESCRIPTION
**Pull request by sdroege**
Continuing https://github.com/rutgersc/m3u8-rs/pull/45

### CODE REVIEW

**Comment by sdroege** [comment]
There's also a problem that if parsing a single item from the metadata attributes fails, all of them seem to be put into the "catch all" hashmap instead of either failing parsing completely or skipping that single item.

**Comment by vagetman** [comment]
```
   To support forward compatibility, when parsing Playlists, clients
   MUST:

   o  ignore any unrecognized tags.

   o  ignore any attribute/value pair with an unrecognized
      AttributeName.

   o  ignore any tag containing an attribute/value pair of type
      enumerated-string whose AttributeName is recognized but whose
      AttributeValue is not recognized, unless the definition of the
      attribute says otherwise.
```
https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-10.txt#section-6.3.1

**Comment by sdroege** [comment]
Yeah that's exactly what I'm on in my local changes :)

**Comment by sdroege** [comment]
@vagetman What's your preference btw for unknown enumerated-string values? Right now I have a separate enum variant for those on all those enums, but we could also simply put those into `unknown_tags` instead.

That's what I'm doing right now for e.g. attributes that require a quoted string but actually got an unquoted string, or the other way around.

**Comment by vagetman** [comment]
> @vagetman What's your preference btw for unknown enumerated-string values? Right now I have a separate enum variant for those on all those enums, but we could also simply put those into `unknown_tags` instead.

This is actually a major point, thanks for asking. The standard calls for ignoring a tag with unrecognized values _by clients_. However, I can see several categories of use cases when this can be a problem for _a_crate_:
* _ingested_ manifests  are "half baked" and need processing before becoming compliant when rendered out
* manifests used for some experimentation by vendors with their devices
* manifests intentionally deviated from the standard for devices that have their own interpretation of such values 
 
Marking the whole tag with an unknown value as unknown will add a major headache to do such work. I believe _a _crate_ should be more flexible than a _client_ and should leave a room to allow such experimentations. 

TLDR; Dropping unknown values into a separate enum is warranted. Making whole tag unknown will add too much hustle on (knowingly) doing anything outside of the ever changing standard. (rfc8216bis10 is the most current one, but stay tuned! it expires on May 12th)
 
> That's what I'm doing right now for e.g. attributes that require a quoted string but actually got an unquoted string, or the other way around.

You're doing a lot of work, thanks!

**Comment by sdroege** [comment]
There are multiple different cases to consider here though
 1. unknown "enumerated string" values for non-mandatory attributes. These can easily be placed into the "leftover" attribute hashmap or into a separate `Other` enum variant like in your PR
 2. unknown "enumerated string" values for mandatory attributes. These can't really be placed into the "leftover" attribute hashmap because they are mandatory, so would have to be in a separate enum variant
 3. string type mismatch: various attributes are defined to be quoted strings, or they're defined to be unquoted strings of a specific format (e.g. a decimal number). We could make all those special enums with a `Other` variant but that seems to make usage unnecessarily complicated (e.g. framerate: do you want to always have to deal with the possibility that it's not a floating point number?)

I think I would prefer at this point to move those unknown "enumerated string" values in any case into a `Other` variant on the existing enum (1/2), but if attributes have a very specific type defined and they're actually something else then I would consider that an error (if mandatory) or have it fall into the "leftover" hashmap (if optional). Or else we would have no plain values anywhere and have to deal with enums everywhere instead.

**Comment by vagetman** [comment]
throwing an error on unknown enumerated string is not making the solution future-proofed. a leftover hashmap sounds reasonable. if a string becomes "known" in a next standard version, but it's not implemented here yet, people still could use it .

**Comment by sdroege** [comment]
That was not the case I was talking about though. Anyway, I'll finish my code and then you can take a look. That's less ambiguous

**Comment by vagetman** [comment]
@sdroege just a friendly ping. :) How is it coming? Need help?

**Comment by sdroege** [comment]
I got distracted by an unexpected amount of work but this is still on my list, no worries! I hope to get to it again next week.

**Comment by sdroege** [comment]
I don't think I'll get back to this in the near future, my to do list is currently overflowing. If you want to take it up, please go ahead, otherwise I'll get back once I have some spare time again.

**Comment by vagetman** [comment]
I've got back to this today and done some work, will commit some code soon.. Just FYI, so we wouldn't bump on each other. 

BTW, @sdroege, I forked off your `attribute-types-wip` branch. Should send a PR on that one first or to `attribute-types` (one on this PR)?

**Comment by sdroege** [comment]
That one should be merged into `attribute-types` and then work be done on that. Thanks for taking this over!

**Comment by vagetman** [comment]
I'm not sure if I need to do this here, but I sent a PR on  the `attribute-types` branch.  https://github.com/sdroege/m3u8-rs/pull/2

**Comment by sdroege** [comment]
@vagetman Good to be merged for you?

**Comment by vagetman** [comment]
woohoo! fantastic news! Sorry for stating the obvious, are we going to publish this as version 5.0.0? Not sure if it's already on somebody's list :)

**Comment by sdroege** [comment]
Sounds like a good idea, I'll try to get that done tomorrow but otherwise someone else has to do it or it has to wait for two weeks.

**Comment by vagetman** [comment]
Hey @rutgersc, can you publish this as v5.0.0 ?

**Comment by rutgersc** [comment]
> Hey @rutgersc, can you publish this as v5.0.0 ?

Done. I was planning to create a PR for it, but accidentally pushed to master. So I've also turned on branch protection, requiring at least 1 approval :)
